gatsby/content/blog/2021/12/2021-12-03-sydent-typing-1.mdx

```diff
@@ -0,0 +1,187 @@
+- we tried to look up a `str` key in a dictionary which mapped `bytes` to `bytes`.
+
+Could a static analysis tool like `mypy` detect these? How much work would it take to do so? Are there other bugs and problems we could spot with it? I was curious to answer these questions and learn more about the tools that Python's typing ecosystem provides.
+
+## Could typing have spotted these problems?
+
+Let's start with the first of question: what can `mypy` detect?
+
+### The missing `await`
+
+Instead of writing `x = await foo()`, we simply had `x = foo()` and didn't then go on to `await x`. Mypy doesn't have means to detect this at present. There was interest in [this issue](https://github.com/python/mypy/issues/2499) on such a feature, with related threads [here](https://github.com/python/mypy/issues/5597) and [here](https://github.com/python/mypy/issues/5650).
+
+Are there other opportunities to spot the error? Here's the relevant bit of source code from [before the fix](https://github.com/matrix-org/sydent/blob/2e3b7576b17e92080319e08c0c0ebf566935636c/sydent/http/servlets/msisdnservlet.py#L93-L98).
+
+```python
+            sid = self.sydent.validators.msisdn.requestToken(
+                phone_number_object, clientSecret, sendAttempt, brand
+            )
+            resp = {
+                "success": True,
+                "sid": str(sid),
+                "msisdn": msisdn,
+                "intl_fmt": intl_fmt,
+            }
+```
+
+The call to `requestToken` produces a value of type `Awaitable[int]`. If we tried to assign that to an expression of type `int` we'd get an error that mypy can spot.
+
+```
+$ cat example.py
+async def foo() -> int:
+    return 1
+
+async def bar():
+    x = foo()      # no error
+    y: int = foo() # error: rhs is Awaitable[int], but lhs expects int
+
+$ mypy --check-untyped-defs example.py
+example.py:6: error: Incompatible types in assignment (expression has type "Coroutine[Any, Any, int]", variable has type "int")
+Found 1 error in 1 file (checked 1 source file)
+```
+
+Note that we have to specifically ask mypy to typecheck the body of `bar` by passing [`--check-untyped-defs`](https://mypy.readthedocs.io/en/stable/command_line.html#cmdoption-mypy-check-untyped-defs); by default, `mypy` will only typecheck annotated code.
+
+We might also have been able to detect the error by looking at how we used `sid`. Unfortunately, the only use of was in a conversion `str(sid)`, which is a perfectly type-safe call for both `sid: int` and `sid: Awaitable[int]`. But let's put that aside for a second—suppose we added `"sid": sid` directly into the `resp` dictionary. Could mypy have spotted _that_?
+
+Unfortunately not. Because `resp` has no annotation, we have to rely on how it's used to assign it a type. And it's only really used as a return value. What's the function's annotation?
```

Comment:
I got very slightly confused at "What's the function's annotation?" Maybe add something clarifying that you are talking about the function that returns `resp`, i.e. something like "What's the annotation of the function that returns `resp`?"

Reply:
0854de6 tries to make this clearer---what do you think?

Reply:
Perfect!